### ISSUE TITLE
Jenkins Jobs Hang with 4.11 but not with 4.10.2

### ISSUE THREAD

**Issue by loren138** [webdriverio/webdriverio#2624]
## The problem

When running our Jenkins file with Web Driver 4.11 we get stuck in a hang with the `cat` command open.  Things work fine with 4.10.2.

## Environment

* WebdriverIO version: 4.11/4.10.2
* Node.js version: 9.6
* [Standalone mode or wdio testrunner](http://webdriver.io/guide/getstarted/modes.html): standalone
* if wdio testrunner, running synchronous or asynchronous tests:
* Additional wdio packages used (if applicable):

## Details

Describe in more detail the problem you have been experiencing, if necessary.

I'm using the Docker image: circle-ci/node:9-stretch-browsers in Jenkins.  (Note: Both versions of webdriver work in the docker image, only when it is mounted on Jenkins do the issues surface.  The new version does seem to hang in the base docker image if it's used with start-server-and-test.)

```

    "e2e:ci": "E2E_MODE=ci BASE_URL=http://localhost:5000 ./node_modules/.bin/wdio ./e2e/wdio.conf.js",
    "e2e:start-ci": "start-server-and-test start:server http://localhost:5000 e2e:ci",
```

When running `./node_modules/.bin/wdio --version`, it correctly returns the version in both cases.

Running `./node_modules/webdriverio/bin/wdio ./e2e/wdio.conf.js` hangs.  Our config file starts with

```
let capabilities = [];
let services = [];
let sauceUser = '';
let sauceKey = '';
console.log('config');
```

The `console.log` is never printed.

I know some changes were made with `argv` and I'm wondering if https://github.com/webdriverio/webdriverio/compare/v4.10.2...v4.11.0#diff-f8849ae9f0614ba9464fdd14c2d98dc0L437 might be related and causing it to wait for input off stdin for some reason.

I'm happy to help debug.  No logs seem to be generated before the hang.

**Comment by christian-bromann** [comment]
@loren138 can you provide some logs or can you help me debug this issue? I am not sure if the argv thing is causing this issue. It could be more related to https://github.com/webdriverio/webdriverio/compare/v4.10.2...v4.11.0#diff-f8849ae9f0614ba9464fdd14c2d98dc0L437

**Comment by loren138** [comment]
@christian-bromann  I didn't see any logs or output produced by webdriver before the hang.  Should there have been something produced in logs if it was getting stuck at one of these locations?  If there should be something somewhere, I can dig around to find it.

I've never tried replacing a node-module with something else, but I'm happy to make a bug branch in our project and run some custom code for web-driver on Jenkins to help with debugging.  I think I'd just need to commit the compiled bin file and execute it with our config file or is it more complicated than that?

**Comment by christian-bromann** [comment]
> I didn't see any logs or output produced by webdriver before the hang.

Not even the one that shows which NPM command was run?

> Should there have been something produced in logs if it was getting stuck at one of these locations?

Actually if you have a console log in your wdio.conf it should be displayed in your stdout

> I forgot I should be able to depend on a commit sha from github rather than using the npm registry which should be fairly easy.

Not sure, doing that probably doesn't give you the compiled version, so you would need to build it as part of your prebuild step

**Comment by akloeber** [comment]
We are observing a very similar behavior. Starting with version `4.11` wdio hangs when executed from IntelliJ IDEA / Webstorm.

The IntelliJ run configuration looks like this:
```
Working Directory: <the project directory>
JavaScript file: node_modules/.bin/wdio
Application parameters: wdio.conf.js
```

This is equivalent to this run configuration file (`.idea/runConfigurations/Run_wdio.xml`):
```xml
<component name="ProjectRunConfigurationManager">
  <configuration default="false" name="Run wdio" type="NodeJSConfigurationType" factoryName="Node.js" activateToolWindowBeforeRun="false" application-parameters="wdio.conf.js" path-to-js-file="node_modules/.bin/wdio" working-dir="$PROJECT_DIR$">
    <method />
  </configuration>
</component>
```

The only output I see in the _"Run"_-window is this:
```
/Users/me/.nvm/versions/node/v8.9.4/bin/node /Users/me/myproject/node_modules/.bin/wdio wdio.conf.js
```
Then the process seems to be stuck forever. On the other hand, if I run the above command from a shell it works as expected.

I also put a `throw new Error('stop')` at the top of `wdio.conf.js` but even that has no effect when executed from IntelliJ and wdio hangs so I expect that `wdio.conf.js` is not read at all. From a shell I see the expected stacktrace provoked by this error:
```
$ Failed loading configuration file: /Users/me/myproject/wdio.conf.js
/Users/me/myproject/node_modules/webdriverio/build/lib/utils/ConfigParser.js:193
                throw e;
                ^

Error: stop
    at Object.<anonymous> (/Users/me/myproject/wdio.conf.js:5:7)
    at ...
```
(The same applies for any console output emitted in `wdio.conf.js`)

If I replace the application parameter `wdio.conf.js` with `-h` or `-v` the help resp. version is printed even if started from IntelliJ. If I switch back to version `4.10.2` everything is fine and tests are executed normally.

**Comment by christian-bromann** [comment]
Interesting, I assume it is due to the execArgv change but I have no idea why id doesn't print anything.

**Comment by loren138** [comment]
Is there a way we can test a rollback of `execArgv` and see if it fixes things?  Or is there more information you need from us?

**Comment by christian-bromann** [comment]
well I thought I improved `execArgv` for the better but I am still not 100% certain if this is really the root cause.

**Comment by WillBrock** [comment]
It could be from #2617.

@akloeber Can you check in cli.js if the launch() function is ever being called when you run it from IntelliJ IDEA / Webstorm?

**Comment by akloeber** [comment]
@WillBrock `launch()` is not called at all when executed from within IntelliJ. It does get called when executing from shell.

**Comment by akloeber** [comment]
@WillBrock The difference is that `process.stdin.isTTY` resolves to `false` when started from IntelliJ and then it keeps waiting for `stdin` to emit `end` which never happens. In the shell `process.stdin.isTTY` is `true` and `launch()` is called directly.

**Comment by akloeber** [comment]
Looks like this has been (re)introduced in #2617 that itself reverted #2477 which might have been added for exactly this scenario where a node process is spawned.

**Comment by christian-bromann** [comment]
What would be the appropriate fix for this?

**Comment by akloeber** [comment]
The options provided by @jasonk in #2617 are not directly applicable (at least in my case) as IntelliJ does actually start the process directly and I have no influence on how it does so (note: there is no intermedia node process that itself spawns `wdio`).

I ran a little test script through IntelliJ:
```
if (process.stdin.isTTY) {
  console.log('process.stdin.isTTY is TRUE');
} else {
  console.log('process.stdin.isTTY is FALSE');
  let stdinData = '';

  const stdin = process.openStdin();
  stdin.on('data', function (data) {
    stdinData += data;
    console.log('DATA', data);
  });
  stdin.on('end', function () {
    console.log('END', stdinData);
  });
}
```

When the script is running I'm able to enter data into the embedded Run-view of IntelliJ which is recognised and logged by the test script. So the Run-view seems to behave "like a TTY" in a way that input and output get passed between the node and the IntelliJ process but `process.stdin.isTTY` is `false`.

Right now I do not see a reliable way of detecting this from inside wdio. It might be an option to add a CLI parameter to wdio that explicitely enables/disabled TTY. Another option would be to add an alternate start script next to `cli.js` for "embedded" execution (there are already a bunch of others whose purpose is not clear to me).

Otherwise each project would have to write a custom wrapper script that interacts with wdio's `Launcher` as documented at http://webdriver.io/guide/testrunner/gettingstarted.html#Run-the-test-runner-programmatically. Nevertheless there should be a section in the documentation that provides instructions on how to run wdio embedded in other tools as otherwise a lot of users will have issues like this.

**Comment by WillBrock** [comment]
I can't really think of any good solution for this. Adding some type of cli parameter might be the way to go.

**Comment by loren138** [comment]
It took almost 4 months between #2477 and #2617 and only 3 days after #2617 for this complaint so it might make more sense to have a flag to read from stdin rather than a flag for cli as I think most people would expect cli mode out of a test runner rather than a read from stdin mode, but that is a breaking change based on the documentation so if you think a flag for cli makes more since, `--cli` sounds good to me.

**Comment by smetnana** [comment]
We have the same problem and can confirm what has been said previously. We have a Jenkins Job that creates a Docker container (node:latest) to run some mocha/wdio test scripts. The job gets stuck as soon as wdio is called, with little useful output:

```
[Pipeline] sh

[dev-test (master)] Running shell script
+ npm run ukcs -- --host selenium

> meera-mdr-test@1.0.0 ukcs /var/jenkins_home/workspace/dev-test (master)
> wdio wdio.conf --suite ukcs "--host" "selenium"
```

At this point the process is stuck without and further output. Our setup has too many moving parts to easily reproduce, so I won't go into further details here. 

However, I can login to the stuck Docker container using a terminal connection and run the exact same command and it works fine. The main difference is that Jenkins does not have a terminal/TTY.

I can confirm that downgrade to `4.10.2` fixes the problem.

**Comment by akloeber** [comment]
@WillBrock @christian-bromann I think `wdio` should behave like other unix tools, e.g. `grep` that do not have any internal magic to detect whether to read from an input file (given as a parameter) or from stdin. The behavior is only determined by the parameters given to the command.

From `grep`'s manpage:

> grep searches the named input FILEs (or standard input if no files are named, or if a single hyphen-minus (-) is given as file name)

A short test confirms that if input is provided via a file parameter and a pipe to stdin at the same time the file parameter has the precedence:
```bash
$ echo foo.from.pipe > pipe
$ echo foo.from.file > file
$ cat pipe | grep foo file
foo.from.file
```

This also enforces @loren138's point that CLI mode should be the default rather than reading from stdin.

**Comment by christian-bromann** [comment]
@akloeber I agree with you. That sounds like a good solution!

**Comment by mkj28** [comment]
+1 from me to get this resolved. Prevents us from upgrading beyond 4.10.2

**Comment by emilmadry** [comment]
+1 from me. Webstorm npm run doesn't work while writing same stuff in terminal is working.
4.10.1 and 4.10.2 is working fine, 4.11 and 4.12 not

**Comment by Meeeeow** [comment]
Same here.

It should't work like a magic, at least it should't wait for stdin while there are configure file path in arguments. 
It would be nice to work like other awesome tools, use `wdio --config -` | `wdio -` to tell command read configure from stdin

**Comment by theansaname** [comment]
:+1: seeing this on my windows install.

**Comment by askkaz** [comment]
+1 Seeing this with gitbash on Windows 10. Reverting to 4.10.2 works.

**Comment by christian-bromann** [comment]
@loren138 feel free to provide a PR. I am not working on a fix for this.

**Comment by akloeber** [comment]
@loren138 I'm not working on a pull request.
I would also recommend to check for a specific parameter (e.g. `"-"`, i.e. `wdio - --watch wdio.conf.js`) instead of the absence of a parameter as there are at least 2 parameters that could specify which tests to run: `--suite` and `--spec`. Furthemore these parameters are optional as they can be omitted in favor of the `specs`-property in `wdio.conf.js`. In this case `wdio` could be invoked without any parameter at all. So checking for a specific flag to enable reading of specs from stdin seems to be the best approach.

**Comment by prakashomp** [comment]
I am using 4.12.0 In parallel execution jobs are hanging with error, with older versions it didn't happen.
 Error: listen EADDRINUSE :::56936
12:46:56     at Object.exports._errnoException (util.js:1026:11)
12:46:56     at exports._exceptionWithHostPort (util.js:1049:20)
12:46:56     at Server._listen2 (net.js:1257:14)
12:46:56     at listen (net.js:1293:10)
12:46:56     at Server.listen (net.js:1389:5)
every time showing different port number, for error message.

**Comment by ezwanharun** [comment]
Any updates on this issue? I also having the similar issue with bamboo , webdriverio 4.12.0 where is stuck and hang at onPrepare hook.

**Comment by kapilchokhawala** [comment]
+1 Seeing this with webdriverio 4.12.0. Blocking me.

**Comment by loren138** [comment]
A pull has been merged to fix it, they just haven't made a new release since then so it should be fixed when the next version is released.

**Comment by tnypxl** [comment]
@christian-bromann What is blocking a 4.12.1 release that would fix this issue?

**Comment by degerlac** [comment]
This bug is biting us too in IntelliJ. Can't wait for this fix to be released.

**Comment by loren138** [comment]
Released in `4.13.0` (note there is now a `4.13.1`)

**Comment by KGBemployee** [comment]
FIx works with Intellij and release 4.13.1
